docs/proposals/signal-api.md

```diff
@@ -0,0 +1,194 @@
+that behavior in the PodSpec and then use the *bump* term when calling the notification API resource.
+
+Once me make agreement on the API part, the document will be extended with the implementation details.
+
+One thing that needs to be noticed is that restarting scenarios are not a goad of this proposal. The only goal here
+is to send single signals, no more, no less. More complicated restart scenarios (like sending SIGTERM, then waiting
+for graceful termination, then sending SIGKILL after reaching timeout) may need another resource in Kubernetes API,
+which will not expose any POSIX terminology to the user.
+
+#### Containers and pods
+
+Despite the fact that Docker supports shared PID namespaces, Kubernetes doesn't, which means that signals can be
+sent only to concrete containers, not pods.
+
+## Motivation
+
+Many Kubernetes users and operators would like to send custom signals to the processes running in containers. That
+may be motivated both by default behaviour that signals are bringing and the custom signal handling which the
+program is doing.
+
+The more concrete use-cases are described below.
+
+## Use-Cases
+
+#### Stopping applications
+
+Currently there is no way to tell applications in containers in pods to stop without deleting a pod. Assuming that
+a pod has *Always* or *OnFailure* restartPolicy, we can make a possibility of manual restart for pods by sending the
+SIGTERM signal.
+
+#### Force stopping applications
+
+Similarly, we can make a possibility of manual force restart by sending the SIGKILL signal.
+
+#### Re-reading configuration
+
+Many applications follow the convention to re-read configuration files when handling the SIGHUP signal. Ability
+to send SIGHUP signal will allow us to reconfigure pods without updating their definition or removing them. Also,
+we can consider using this API in the watch mechanism in kubelet to reconfigure a pod automatically on the ConfigMap
+change.
+
+## Analysis
+
+The system needs to be able to:
+
+1. Model correctly which signals are supported
+2. Allow to define what signals can be used for each pod (and whether they can be used at all)
+3. Provide custom names for the chosen signals in pod
+
+## Proposed design
+
+Signal API will be a part of pod API. There will be no separate API group for that.
+
+#### Schema
+
+The proposed schema is as follow.
+
+There will be a subresource of `pod` called `NotifyREST`, defined like:
+
+```go
+type NotifyREST struct {
+	Store       *registry.Store
+	KubeletConn client.ConnectionInfoGetter
+}
+```
+
+There will be a type for options:
+
+```go
+type PodNotifyOptions struct {
+	unversioned TypeMeta `json:",inline"`
+	Container string `json:"container,omitempty" protobuf:"bytes,1,opt,name=container"`
+	Name string `json:"name" protobuf:"bytes,2,opt,name=name"`
+}
+```
+
+Finally, here will be a type Notification being a part of the PodSpec:
```

Comment:
Putting this on PodSpec is a little weird, because it forces you to carry the `Container` parameter in your options struct or in the spec.  This could be defined on a `type Container`, which implies the target container, and more closely matches Liveness and Readiness.

That said, if we add something like HTTPPost as an option, that doesn't really have a container binding.

What's least awful?

```
Pod.spec.notifications:
  - name: bump
    type: Signal
    signal: SIGHUP
    container: my-server
  - name: post
    type: HTTPPost
    httpPost:
        port: 80
        path: /notify
        body: Boo
```

or

```
Pod.spec.containers:
  - name: my-server
    notifications:
      - name: bump
        type: Signal
        signal: SIGHUP
      - name: post
        type: HTTPPost
        httpPost:
            port: 80
            path: /notify
            body: Boo
```